### PR TITLE
fetch the last not empty coin balance records

### PR DESCRIPTION
**Pull request by ayrat555**
fixes https://github.com/poanetwork/blockscout/issues/2535
## Motivation

We were fetching coin balance records for coin balance chart in the period of 90 days. This PR fetches last 90 not empty token balance records

## Changelog

- fetch the last not empty coin balance records

### CODE REVIEW

File: apps/block_scout_web/assets/js/lib/coin_balance_history_chart.js [hunk 1]
```diff
@@ -18,6 +18,14 @@ export function createCoinBalanceHistoryChart (el) {
           y: balance.value
         }))
 
+      var stepSize = 3
+
+      if (data.length > 2) {
+        var diff = Math.abs(new Date(data[0].date) - new Date(data[1].date))
+        var periodInDays = diff / (1000 * 60 * 60 * 24)
+        console.log(periodInDays)
```
Comment (vbaranov) [inline apps/block_scout_web/assets/js/lib/coin_balance_history_chart.js]: Please remove `console.log`

File: apps/explorer/lib/explorer/chain.ex [hunk 2]
```diff
@@ -2965,6 +2965,7 @@ defmodule Explorer.Chain do
     |> CoinBalance.fetch_coin_balances(paging_options)
     |> page_coin_balances(paging_options)
     |> Repo.all()
+    |> Enum.dedup()
```
Comment (pasqu4le) [inline apps/explorer/lib/explorer/chain.ex]: Why is this needed? Could `CoinBalance.fetch_coin_balances` return exact duplicates?

File: apps/explorer/lib/explorer/chain/address/coin_balance.ex [hunk 3]
```diff
@@ -89,16 +89,32 @@ defmodule Explorer.Chain.Address.CoinBalance do
...
     |> where([cb], cb.address_hash == ^address_hash)
-    |> where([cb, b], b.timestamp >= fragment("date_trunc('day', now()) - interval '90 days'"))
+    |> limit_time_interval(block_timestamp)
     |> group_by([cb, b], fragment("date_trunc('day', ?)", b.timestamp))
     |> order_by([cb, b], fragment("date_trunc('day', ?)", b.timestamp))
     |> select([cb, b], %{date: type(fragment("date_trunc('day', ?)", b.timestamp), :date), value: max(cb.value)})
   end
 
+  def limit_time_interval(query, nil) do
+    query |> where([cb, b], b.timestamp >= fragment("date_trunc('day', now()) - interval '90 days'"))
```
Comment (pasqu4le) [inline apps/explorer/lib/explorer/chain/address/coin_balance.ex]: AFAIU this does the same as the other `limit_time_interval` definition, but providing today as the default value.
To avoid duplication, could this not be defined in term of the other definition?

Comment (ayrat555) [reply apps/explorer/lib/explorer/chain/address/coin_balance.ex]: they differ in fragmenst. but fragments can only be defined inside of query so the can not be abstracted to the common method

File: apps/explorer/lib/explorer/chain.ex (hunk 2 above)
Comment (ayrat555) [reply apps/explorer/lib/explorer/chain.ex]: you're right. fixed

File: apps/block_scout_web/assets/js/lib/coin_balance_history_chart.js (hunk 1 above)
Comment (ayrat555) [reply apps/block_scout_web/assets/js/lib/coin_balance_history_chart.js]: done

**Comment by ayrat555** [comment]
@vbaranov  I fixed the chart

**Comment by vbaranov** [comment]
@ayrat555 

`/address/0xebd37b256563e30c6f9289a8e2702f0852880833/coin_balances`

Screenshot 2019-08-12 at 17 59 51

**Comment by ayrat555** [comment]
@vbaranov I fixed it. the issue was in chart steps

**Comment by ayrat555** [comment]
@vbaranov the issue with duplicate entries is caused by duplicate entries in the DB. Can you please create an issue for it? I'll resolve it in future PRs. I think the issue with the chart is completely fixed by this PR

**Comment by vbaranov** [comment]
@ayrat555 tests are broken

**Comment by ayrat555** [comment]
@vbaranov the test is fixed
